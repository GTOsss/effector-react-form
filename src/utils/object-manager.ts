import stringToPath from 'lodash.topath';

/**
 * @param {object} state Redux state
 * @param path
 * @param removeEmpty
 * @param inDeep should be false for inlineMap
 * @param index
 * @returns {object} State
 */
export const deleteIn = (
  state: Record<string, any>,
  path: Array<string> | string,
  removeEmpty = false,
  inDeep = true, // false for inlineMap
  index = 0,
): any => {
  let pathArray = index === 0 ? stringToPath(path) : path;
  if (!inDeep) {
    pathArray = [path];
  }

  const currentKey = pathArray[index];

  if (!state) {
    return state;
  }

  const isEndPoint = pathArray.length - 1 === index;
  const isArray = Array.isArray(state);
  const isObject = !isArray && typeof state === 'object';
  // @ts-ignore
  const newState = isArray ? [...state] : { ...state };

  if (isObject) {
    if (isEndPoint) {
      // tslint:disable-next-line:no-dynamic-delete
      delete newState[currentKey];
      return newState;
    } else {
      const result = deleteIn(newState[currentKey], pathArray, removeEmpty, inDeep, index + 1);
      const isRemoveEmpty = !result || (removeEmpty && (!result || !Object.keys(result).length));
      if (isRemoveEmpty) {
        // tslint:disable-next-line:no-dynamic-delete
        delete newState[currentKey];
      } else {
        newState[currentKey] = result;
      }
      return newState;
    }
  } else if (isArray) {
    if (isEndPoint) {
      return newState.filter((_, i) => Number(currentKey) !== i);
    } else {
      const result = deleteIn(newState[currentKey], pathArray, removeEmpty, inDeep, index + 1);
      const isRemoveEmpty = !result || (removeEmpty && (!result || !Object.keys(result).length));
      if (isRemoveEmpty) {
        return newState.filter((_, i) => Number(currentKey) !== i);
      } else {
        newState[currentKey] = result;
      }
      return newState;
    }
  }
};

/**
 *
 * @param {object} state Object will be expand
 * @param {string} path Path in object
 * @param {*} value Value will be add
 * @param {number} pathIndex Not require
 * @returns {object} State
 */
export const setIn = (state, path, value, pathIndex = 0) => {
  const pathArray = pathIndex === 0 ? stringToPath(path) : path;

  if (pathIndex >= pathArray.length) {
    return value;
  }

  const first = pathArray[pathIndex];
  const firstState = state && (Array.isArray(state) ? state[Number(first)] : state[first]);
  const next = setIn(firstState, pathArray, value, pathIndex + 1);

  if (!state) {
    if (isNaN(first)) {
      return { [first]: next };
    }
    const initialized = [];
    // @ts-ignore
    initialized[parseInt(first, 10)] = next;
    return initialized;
  }

  if (Array.isArray(state)) {
    // @ts-ignore
    const copy = [].concat(state);
    // @ts-ignore
    copy[parseInt(first, 10)] = next;
    return copy;
  }

  return {
    ...state,
    [first]: next,
  };
};

/**
 *
 * @param {object} state Redux state
 * @param {string} field Path to field
 * @param {*?} defaultValue
 * @returns {object} State
 */
export const getIn = (state, field, defaultValue?) => {
  if (!state) {
    return defaultValue;
  }

  const path = stringToPath(field);
  const { length } = path;
  if (!length) {
    return defaultValue;
  }

  let result = state;
  for (let i = 0; i < length && result; i += 1) {
    result = result[path[i]];
  }

  if (!result) {
    return defaultValue;
  }

  return result;
};

export const makeNested = <Result = Record<string, any>>(inlineMap: Record<string, any>): Result =>
  Object.entries(inlineMap).reduce((acc, [key, value]) => setIn(acc, key, value), {} as Result);

const isFieldMeta = (value) => typeof value === 'object' && value._type === 'fieldMeta';

export const removeFromInlineMap = (map: Record<string, any>, key: string) => {
  const nestedMap = deleteIn(makeNested(map), key);
  const nodes = [{ node: nestedMap, path: [] }];
  const newInlineMap = {};

  while (nodes.length) {
    const currentNode = nodes.pop();
    Object.entries(currentNode.node).forEach(([k, v]) => {
      if (isFieldMeta(v)) {
        newInlineMap[currentNode.path.concat(k).join('.')] = v;
      } else {
        nodes.push({
          node: v,
          path: currentNode.path.concat(k),
        });
      }
    });
  }

  return newInlineMap;
};

export const makeConsistentKey = (key) => stringToPath(key).join('.');
