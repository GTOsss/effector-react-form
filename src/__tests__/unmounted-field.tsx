import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';
import {createStore, createEvent} from 'effector';
import {useStore} from 'effector-react';
import {useForm} from '../';
import {Controller, FormState} from '../../index';
import {setIn, removeFromInlineMap} from '../utils/object-manager';

const renderForm = () => {
  const validateRequired = (value) => value ? undefined : 'Field is required';

  type Values = {
    fields: Array<{value: string, id: number | string}>,
  };

  const removeFirstElement = createEvent();

  const $fieldsInline = createStore({});
  const $values = createStore<Values>({
    fields: [
      {value: '', id: 1},
      {value: '', id: 2},
    ],
  });

  $fieldsInline.on(removeFirstElement, (inlineState) => removeFromInlineMap(inlineState, 'fields[0]'));
  $values.on(removeFirstElement, (state) => setIn(state, 'fields', state.fields.slice(1)));

  const $form = createStore<FormState>({
    submitted: false,
    hasError: false,
    hasOuterError: false,
  });

  const $errorsInline = createStore({});

  interface InputProps {
    controller: Controller,
    label: string,
  }

  const Input: React.FC<InputProps> = ({
    controller,
    label,
  }) => {
    const {input, fieldState, form, error} = controller();

    const isShowError = form.submitted || form.hasOuterError || fieldState.blurred;

    return (
      <div role="wrapper-for-input" className="input-wrap">
        <label>{label}</label>
        <input
          {...input}
          value={input.value || ''}
          role="textbox"
          className="input"
          placeholder={label}
        />
        {isShowError && error && (<span>{error}</span>)}
      </div>
    );
  };

  const SimpleForm = () => {
    const {handleSubmit, controller} = useForm<Values>({$values, $fieldsInline, $form, $errorsInline, onSubmit: () => {}});

    const {fields} = useStore($values);

    return (
      <form onSubmit={handleSubmit}>
        {fields.map((el, i) => (
          <Input
            key={el.id}
            controller={controller({name: `fields[${i}].value`, validate: validateRequired})}
            label={`value â„– ${i}`}
          />
        ))}
        <button type="submit">submit</button>
        <button
          type="button"
          onClick={() => removeFirstElement()}
        >
          remove first element
        </button>
      </form>
    );
  };

  render(<SimpleForm />);

  return {
    $errorsInline,
    $values,
    $form,
  };
};

describe('UnmountField', () => {
  test('render fields', () => {
    renderForm();
    const inputs = screen.getAllByRole('textbox');
    expect(inputs).toMatchSnapshot();
  });

  test('render fields after removed', () => {
    renderForm();
    const remove = screen.getByText('remove first element');
    fireEvent.click(remove);
    const inputs = screen.getAllByRole('textbox');
    expect(inputs).toMatchSnapshot();
  });

  test('render fields after submit and removed', () => {
    renderForm();
    const submit = screen.getByText('submit');
    fireEvent.click(submit);
    const remove = screen.getByText('remove first element');
    fireEvent.click(remove);
    const inputs = screen.getAllByRole('wrapper-for-input');
    expect(inputs).toMatchSnapshot();
  });

  test('$form after submit and removed', () => {
    const {$form} = renderForm();
    const submit = screen.getByText('submit');
    fireEvent.click(submit);
    const remove = screen.getByText('remove first element');
    fireEvent.click(remove);
    expect($form.getState()).toMatchSnapshot();
  });

  test('$errorsInline after submit > removed', () => {
    const {$errorsInline} = renderForm();
    const submit = screen.getByText('submit');
    fireEvent.click(submit);
    const remove = screen.getByText('remove first element');
    fireEvent.click(remove);
    expect($errorsInline.getState()).toMatchSnapshot();
  });

  test('$errorsInline after submit > removed > submit', () => {
    const {$errorsInline} = renderForm();
    const submit = screen.getByText('submit');
    fireEvent.click(submit);
    const remove = screen.getByText('remove first element');
    fireEvent.click(remove);
    fireEvent.click(submit);
    expect($errorsInline.getState()).toMatchSnapshot();
  });

  test('$values after change 1,2 > removed', () => {
    const {$values} = renderForm();
    const inputs = screen.getAllByRole('textbox');
    fireEvent.change(inputs[0], {target: {value: 'value 1'}});
    fireEvent.change(inputs[1], {target: {value: 'value 2'}});
    const remove = screen.getByText('remove first element');
    fireEvent.click(remove);
    expect($values.getState()).toMatchSnapshot();
  });

  test('$values after removed > onChange("test")', () => {
    const {$values} = renderForm();
    const submit = screen.getByText('submit');
    fireEvent.click(submit);
    const remove = screen.getByText('remove first element');
    fireEvent.click(remove);
    const input = screen.getByRole('textbox');
    fireEvent.change(input, {target: {value: 'test'}});
    expect($values.getState()).toMatchSnapshot();
  });
});


