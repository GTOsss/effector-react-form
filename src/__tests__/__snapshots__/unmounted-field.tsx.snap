// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnmountField $errorsInline after submit > removed > submit 1`] = `
Object {
  "fields[0].value": "Field is required",
}
`;

exports[`UnmountField $errorsInline after submit > removed 1`] = `
Object {
  "fields[0].value": "Field is required",
}
`;

exports[`UnmountField $form after submit and removed 1`] = `
Object {
  "hasError": true,
  "hasOuterError": false,
  "submitted": true,
}
`;

exports[`UnmountField $values after removed > onChange("test") 1`] = `
Object {
  "fields": Array [
    Object {
      "id": 2,
      "value": "test",
    },
  ],
}
`;

exports[`UnmountField render fields 1`] = `
Array [
  <input
    class="input"
    name="fields[0].value"
    placeholder="value № 0"
    role="textbox"
    value=""
  />,
  <input
    class="input"
    name="fields[1].value"
    placeholder="value № 1"
    role="textbox"
    value=""
  />,
]
`;

exports[`UnmountField render fields after removed 1`] = `
Array [
  <input
    class="input"
    name="fields[0].value"
    placeholder="value № 0"
    role="textbox"
    value=""
  />,
]
`;

exports[`UnmountField render fields after submit and removed 1`] = `
Array [
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      value № 0
    </label>
    <input
      class="input"
      name="fields[0].value"
      placeholder="value № 0"
      role="textbox"
      value=""
    />
    <span>
      Field is required
    </span>
  </div>,
]
`;
